castai/resource_workload_scaling_policy.go

```diff
@@ -0,0 +1,341 @@
+			Memory:           toResourcePolicies(d.Get("memory").([]interface{})[0].(map[string]interface{})),
+		},
+	}
+
+	resp, err := client.WorkloadOptimizationAPIUpdateWorkloadScalingPolicyWithResponse(ctx, clusterID, d.Id(), req)
+	if checkErr := sdk.CheckOKResponse(resp, err); checkErr != nil {
+		return diag.FromErr(checkErr)
+	}
+
+	return resourceWorkloadScalingPolicyRead(ctx, d, meta)
+}
+
+func resourceWorkloadScalingPolicyDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+	clusterID := d.Get(FieldClusterID).(string)
+
+	resp, err := client.WorkloadOptimizationAPIGetWorkloadScalingPolicyWithResponse(ctx, clusterID, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if resp.StatusCode() == http.StatusNotFound {
+		log.Printf("[DEBUG] Scaling policy (%s) not found, skipping delete", d.Id())
+		return nil
+	}
+	if err := sdk.StatusOk(resp); err != nil {
+		return diag.FromErr(err)
+	}
+
+	delResp, err := client.WorkloadOptimizationAPIDeleteWorkloadScalingPolicyWithResponse(ctx, clusterID, d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if delResp.StatusCode() == http.StatusBadRequest {
+		log.Printf("[WARN] Scaling policy has active workloads (%s) and can't be deleted, removing from state", d.Id())
+		return nil
+	}
+	if err := sdk.StatusOk(delResp); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func resourceWorkloadScalingPolicyDiff(_ context.Context, d *schema.ResourceDiff, _ interface{}) error {
+	// Since tf doesn't support cross field validation, doing it here.
+	cpu := toResourcePolicies(d.Get("cpu").([]interface{})[0].(map[string]interface{}))
+	memory := toResourcePolicies(d.Get("memory").([]interface{})[0].(map[string]interface{}))
+
+	if err := validateArgs(cpu, "cpu"); err != nil {
+		return err
+	}
+	return validateArgs(memory, "memory")
+}
+
+func validateArgs(r sdk.WorkloadoptimizationV1ResourcePolicies, res string) error {
+	if r.Function == "QUANTILE" && len(r.Args) == 0 {
+		return fmt.Errorf("field %q: QUANTILE function requires args to be provided", res)
+	}
+	if r.Function == "MAX" && len(r.Args) > 0 {
+		return fmt.Errorf("field %q: MAX function doesn't accept any args", res)
+	}
+	return nil
+}
+
+func workloadScalingPolicyImporter(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	ids := strings.Split(d.Id(), "/")
+	if len(ids) != 2 || ids[0] == "" || ids[1] == "" {
+		return nil, fmt.Errorf("expected import id with format: <cluster_id>/<scaling_policy name or id>, got: %q", d.Id())
+	}
+
+	clusterID, id := ids[0], ids[1]
+	if err := d.Set(FieldClusterID, clusterID); err != nil {
+		return nil, fmt.Errorf("setting cluster id: %w", err)
+	}
+	d.SetId(id)
+
+	// Return if scaling policy ID provided.
+	if _, err := uuid.Parse(id); err == nil {
+		return []*schema.ResourceData{d}, nil
```

Comment:
shouldn't we check if it exists?

Reply:
you mean scaling policy? it will go through read path and will return err if it doesn't exist